include/deal.II/lac/scalapack.h

```diff
@@ -0,0 +1,516 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_scalapack_h
+#define dealii_scalapack_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_SCALAPACK
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/lapack_support.h>
+#include <deal.II/base/mpi.h>
+#include <mpi.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+// Forward declaration of class ScaLAPACKMatrix for ProcessGrid
+template <typename NumberType> class ScaLAPACKMatrix;
+
+
+/**
+ * A class taking care of setting up the two dimensional process grid.
+ * For example an MPI communicator with 5 cores can be arraged into a
+ * 2x2 grid with 5-th processor being inactive:
+ * @code
+ *      |   0     |   1
+ * -----| ------- |-----
+ * 0    |   P0    |  P1
+ *      |         |
+ * -----| ------- |-----
+ * 1    |   P2    |  P3
+ * @endcode
+ *
+ * A shared pointer to this class is provided to ScaLAPACKMatrix matrices to
+ * perform block-cyclic distribution.
+ *
+ * Note that this class allows to setup a process grid which has fewer
+ * MPI cores than the total number of cores in the communicator.
+ *
+ * @author Benjamin Brands, 2017
+ */
+class ProcessGrid
```

Comment:
Since this class is exclusively about MPI, would it make sense to move it to namespace `Utilities::MPI` in `mpi.h`?

Reply:
not really, it uses ScaLAPACK functions to do initialization and provide some interfaces. IMO it should belong to the same header as the ScaLAPACK class.

Reply:
> But if that's how it is, then why not move the class right into the only class that can currently use it?

because this class needs to be shared in dense parallel matrices. When you do generalized eigenvalue problem or such, then all participating matrices should have the same process grid.

Reply:
> then it should expose functionality that could also be used from other places.

i don't think it's necessary as there is not much one can do with it other than pass it to different ScaLAPACK matrices to initialize them.

Reply:
As most of the ScaLapack routines (with only two exceptions for copying) require distributed matrices using the same BLACS context (BLACS analogue to the MPI communicator), which is a member of the class ProcessGrid, we decided to separate it from the class ScaLAPACKMatrix. In particular, we were thinking about SVD, for which we need at least two matrices to store the right and left singular vectors.

Reply:
The question I'm asking is this:

    "Could this class be used by anything other than ScalapackMatrix?"

If the answer is yes, then there is a design problem in that the class does not actually expose any functionality to any potential user *unless that user is declared `friend`*. That means that any future user will have to be declared a friend as well.

If the answer is no, then there is no reason to have this class in the global namespace -- it could in that case move into some `namespace internal::ScalapackMatrix`, or just be a nested member class within the `ScalapackMatrix` class itself.

Reply:
@bangerth 

> "Could this class be used by anything other than ScalapackMatrix?"

No, it can not.

> Then there is no reason to have this class in the global namespace

Yes, there is. It has to be used in constructor to create several matrices with exactly the same grid partitioning.

Reply:
@bangerth in other words, grid partitioner **has to be shared among matrices**, thus ScaLAPACK stores a shared pointer to this class. I see no problem with this design and frankly don't understand your concerns.

Reply:
I understand that you need to share *objects* of this type between different matrix objects, but that is independent of the question of where the *class* is declared. For example, you could have
```
  class ScalapackMatrix {
    public:
      class ProcessGrid { ... };        // public member class

      ScalapackMatrix (std::shared_ptr<ProcessGrid> &process_grid);  // constructor
  };
```

This way, it's clear that the only place where one would use a `ProcessGrid` object is in connection with a `ScalapackMatrix` object..

Reply:
>> For example, you could have

correct, but this means one would need to 

(i) carry along the template number of ScaLAPACKMatrix into the `ProcessGrid`, ie. `ScaLAPACKMatrix<number>::ProcessGrid` which is not right as Process gird does not care about the number type.

(ii) if one day there are other dense parallel linear algebra, then the `ProcessGrid` could be extended to actually cover more than just `ScaLAPACK`. And maybe be usable in different places.

Reply:
> This way, it's clear that the only place where one would use a ProcessGrid object is in connection with a ScalapackMatrix object..

how about a line in documentation stating that, would this be ok?

Reply:
I think we're at an impasse :-)

@dealii/dealii -- I think we need more eyes with different perspectives on this issue!

Reply:
@bangerth since no one comments so far, here is another reason not do what u suggest: it is perfectly valid to share process grid between matrices of different number types. User might want to do some calculations with float, copy this into double matrix distributed over the same grid and do something else.

Reply:
That's fair, but then I still think that the class should have an interface that makes it usable not just in one other class.

@dealii/dealii -- anyone?

Reply:
I agree with @bangerth Either we make `ProcessorGrid` extensible and we move it out of `ScaLAPACK` or we change it so that it doesn't pollute the global namespace. Personally, I would just move it to a different namespace.